packages/compass-aggregations/src/modules/index.ts

```diff
@@ -248,276 +112,6 @@ const appReducer = combineReducers({
-    collationString,
-    pipeline
-  } = action.restoreState;
-
-  return {
-    // Current state will be mostly preserved (i.e, namespace, isTimeSeries, etc)
-    ...state,
-    // Everything that is stored as a pipeline is applied next
-    id,
-    name,
-    comments,
-    sample,
-    autoPreview,
-    collationString: getCollationStateFromString(collationString),
-    pipeline,
-    // Relevant state that depens on the pipeline state is updated (NB: this
-    // whole thing should be happening in the relevant slice reducers instead,
-    // but changing how this plugin works is way too big of a task for this
-    // particular patch)
-    savedPipeline: {
-      ...state.savedPipeline,
-      isListVisible: false
-    },
-    restorePipeline: {
-      isModalVisible: false,
-      pipelineObjectID: ''
-    }
-  };
-};
-
-/**
- * Handle the pipeline clear.
- *
- * @param {Object} state - The state.
- *
- * @returns {Object} The new state.
- */
-const doClearPipeline = (state: RootState): RootState => ({
-  ...state,
-  pipeline: [],
-  limit: LIMIT_INITIAL_STATE,
-  largeLimit: LARGE_LIMIT_INITIAL_STATE,
-  maxTimeMS: MAX_TIME_MS_INITIAL_STATE,
-  isAtlasDeployed: state.isAtlasDeployed,
-  outResultsFn: state.outResultsFn,
-  savedPipeline: {
-    ...state.savedPipeline,
-    isListVisible: true
-  }
-});
-
-/**
- * Create a new pipeline.
- *
- * @param {Object} state - The state.
- *
- * @returns {Object} The new state.
- */
-const createNewPipeline = (state: RootState): RootState => ({
-  ...INITIAL_STATE,
-  aggregationWorkspaceId: state.aggregationWorkspaceId,
-  appRegistry: state.appRegistry,
-  namespace: state.namespace,
-  env: state.env,
-  isTimeSeries: state.isTimeSeries,
-  isReadonly: state.isReadonly,
-  sourceName: state.sourceName,
-  fields: state.fields,
-  serverVersion: state.serverVersion,
-  dataService: state.dataService,
-  isAtlasDeployed: state.isAtlasDeployed,
-  outResultsFn: state.outResultsFn,
-  inputDocuments: state.inputDocuments
-});
-
-/**
- * Create a cloned pipeline.
- *
- * @param {Object} state - The state.
- *
- * @returns {Object} The new state.
- */
-const createClonedPipeline = (state: RootState): RootState => ({
-  ...state,
-  id: new ObjectId().toHexString(),
-  name: `${state.name} (copy)`,
-  isModified: true
-});
-
-/**
- * The action to state modifier mappings.
- */
-const MAPPINGS = {
-  [RESET]: doReset,
-  [RESTORE_PIPELINE]: doRestorePipeline,
-  [CLEAR_PIPELINE]: doClearPipeline,
-  [NEW_PIPELINE]: createNewPipeline,
-  [CLONE_PIPELINE]: createClonedPipeline
-};
-
-/**
- * The root reducer.
- *
- * @param state - The state.
- * @param action - The action.
- *
- * @returns The new state.
- */
-const rootReducer: Reducer<RootState, AnyAction> = (state, action): RootState => {
-  const fn = (MAPPINGS as any)[action.type];
-  return (state && fn) ? fn(state, action) : appReducer(state, action);
-};
+export type RootState = ReturnType<typeof rootReducer>;
 
 export default rootReducer;
```

Comment:
we check for `(action.type === NEW_PIPELINE || action.type === CLEAR_PIPELINE)` at many places. can we add this check to the rootReducer and return the expected state? Like `doReset` which was already there?

Reply:
That's the behavior I'm specifically trying to avoid here. Redux actions are specifically made so that many reducers can react to the same action, this creates an explicit dependency between the slice and action instead of an implicit catch-all that makes it hard to track what exactly and how can change the store. Here's how redux docs explain it:

> We suggest you write independent small reducer functions that are each responsible for updates to a specific slice of state. We call this pattern “reducer composition”. A given action could be handled by all, some, or none of them. This keeps components decoupled from the actual data changes, as one action may affect different parts of the state tree, and there is no need for the component to be aware of this.

https://redux.js.org/faq/actions#is-there-always-a-one-to-one-mapping-between-reducers-and-actions

> combineReducers is deliberately limited to handle a single common use case: updating a state tree that is a plain Javascript object by delegating the work of updating each slice of state to a specific slice reducer.

https://redux.js.org/faq/design-decisions#why-doesnt-combinereducers-include-a-third-argument-with-the-entire-state-when-it-calls-each-reducer

(there is also some further reading in these links if you want to go more in details)

Reply:
Also just to be clear I partially agree that this is super verbose, but mostly caused by the amount of slices we have here that probably could've been smaller.

Also to give you an example of how this existing implementation was making state updates convoluted and hard to track, in current implementation `createNewPipeline` was resetting `isDataLake` and a few other state values to initial state which means that when connected to data lake if you create new pipeline you won't be able to run explain plan anymore